core/src/pretty.rs

```diff
@@ -152,17 +153,197 @@ fn needs_parens_in_type_pos(typ: &Type) -> bool {
     }
 }
 
-pub fn fmt_pretty<'a, T>(value: &T, f: &mut fmt::Formatter) -> fmt::Result
+pub fn fmt_pretty<T>(value: &T, f: &mut fmt::Formatter) -> fmt::Result
 where
-    T: Pretty<'a, pretty::BoxAllocator, ()> + Clone,
+    T: for<'a> Pretty<'a, BoundedAllocator, ()> + Clone,
 {
-    let doc: DocBuilder<_, ()> = value.clone().pretty(&pretty::BoxAllocator);
+    let allocator = BoundedAllocator::default();
+    let doc: DocBuilder<_, ()> = value.clone().pretty(&allocator);
     doc.render_fmt(80, f)
 }
 
-impl<'a, A: Clone + 'a> NickelAllocatorExt<'a, A> for pretty::BoxAllocator {}
+#[derive(Clone, Copy, Debug, Default)]
+struct SizeBound {
+    depth: usize,
+    size: usize,
+}
+
+/// A guard for scoping changes to a `SizeBound`, restoring the old
+/// value when leaving a scope.
+#[derive(Default)]
+struct ShrinkGuard<'a> {
+    bound: Option<&'a Cell<SizeBound>>,
+    old_bound: SizeBound,
+}
+
+impl<'a> Drop for ShrinkGuard<'a> {
+    fn drop(&mut self) {
+        if let Some(bound) = self.bound {
+            bound.set(self.old_bound);
+        }
+    }
+}
+
+/// A pretty-printing allocator that supports rough bounds on the
+/// size of the output.
+///
+/// When a pretty-printed object is too large, it will be abbreviated.
+/// For example, a record will be abbreviated as "{…}".
+///
+/// The bounds are "rough" in that the depth bound only (currently; this might
+/// be extended in the future) constrains the number of nested records: you can
+/// still have deeply nested terms of other kinds. The size bound only constrains
+/// the number of children of nested records. As such, neither constraint gives
+/// precise control over the size of the output.
+pub struct BoundedAllocator {
+    inner: pretty::BoxAllocator,
+    bound: Option<Cell<SizeBound>>,
+}
+
+/// The default `BoundedAllocator` imposes no constraints.
+impl Default for BoundedAllocator {
+    fn default() -> Self {
+        Self {
+            inner: pretty::BoxAllocator,
+            bound: None,
+        }
+    }
+}
+
+impl BoundedAllocator {
+    /// Creates a `BoundedAllocator` with constraints.
+    pub fn bounded(max_depth: usize, max_size: usize) -> Self {
+        Self {
+            inner: pretty::BoxAllocator,
+            bound: Some(Cell::new(SizeBound {
+                depth: max_depth,
+                size: max_size,
+            })),
+        }
+    }
+
+    /// Shrinks this allocator, returning a guard that restores the
+    /// old constraints when it leaves scope.
+    fn shrink(&self, child_size: usize) -> ShrinkGuard<'_> {
+        if let Some(bound) = self.bound.as_ref() {
+            let old = bound.get();
+            bound.set(SizeBound {
+                depth: old.depth - 1,
```

Comment:
Should we defend over underflow ? (nothing prevents from instantiating with `depth=0` and then shrinking)

```suggestion
                depth: old.depth.saturating_sub(1),
```

Reply:
I had this at some point but was worried about the asymmetry between decrementing and then incrementing it back again. I think the underflowing one is actually ok, since it will panic in tests and then behave slightly strangely but deterministically (possibly missing some truncation) in release mode.

Reply:
> I had this at some point but was worried about the asymmetry between decrementing and then incrementing it back again.

But the code is only decrementing it, isn't it? How would this "asymmetry between decrementing and then incrementing it" would manifest in practice, in terms of concrete effect? This isn't really important because we're talking about a parameter that controls pretty printing, so worst case we print too much, but in term of semantics, I find the difference in behavior between debug and release - the latter having a completely unexpected behavior - to be unsatisfactory.

Reply:
The incrementing happens when dropping the guard. But I think this can be solved by storing the old value and then restoring it, like I'm doing with `size`.

Reply:
I see, I thought both were saved and restored. In any case neither wrapping nor incrementing after a saturating sub are correct, but I would still err on the side of wrapping is more incorrect than anything else :stuck_out_tongue: